docs/OptionSettingsItems-input-validation.md

```diff
@@ -0,0 +1,193 @@
+# OptionSettingItems Input Validation Design Document
+_DOTNET-4952_
+
+## Summary
+Need mechanism to indicate an `OptionSettingItem` is required (ie non empty value) or has additional input validations:
+ - less than 64 characters
+ - no whitespace characters
+ - no non-ascii characters
+ - value is from list of acceptable values
+ - is positive number
+ - is number within bounds
+
+Deployment should be blocked if there are any OptionSettingItems with invalid values.
+
+## Proposed Design
+
+Unqiue validation classes will be created and will implement a new interface:
+
+```csharp
+public interface IOptionSettingItemValidator
+{
+    OptionSettingItemValidationResult Validate(object input);
+}
+
+public class OptionSettingItemValidationResult
+{
+    public bool IsValid { get; set; }
+    public string ValidationFailedMessage { get;set; }
+}
+```
+
+OptionSettingItems will then contain a collection of 0 or more `IOptionSettingItemValidator`s:
+
+```csharp
+public class OptionSettingItem
+{
+     public List<IOptionSettingItemValidator> Validators { get; set; }
+}
+```
+
+### Validation
+
+`OptionSettingItem.SetValueOverride` will be updated to validate the override value by passing it to each Validator.  If any validator indicates the value is not valid, an exception is thrown:
+
+```csharp
+public void SetValueOverride(object valueOverride)
+{
+    foreach (var validator in this.Validators)
+    {
+        var result = validator.Validate(valueOverride);
+        if (!result.IsValid)
+            throw new ValidationFailedException
+            {
+                ValidationResult = result
+            };
+    }
+
+    // value is saved
+}
+```
+
+### Serialization of Recipes
+
+`RecipeDefinition`s are stored as json and deseiralized from _*.recipe_ files.  Therefor the Validators must also be Json Serializable.  To faciliate polymorphism, recipe json will need to define the type of the validator.
+
+In the example, below the **Example Setting** item has two Validators, `RequiredValidator` and `RegexValidator`:
+
+```json
+// Minimal Recipe to highlight OptionSettingsItem.Validators
+{  
+  "Name": "Example Recipe",  
+  "OptionSettings": [
+    {      
+      "Name": "Example Setting",
+      "Type": "String",
+      "DefaultValue": null,
+      "Validators": [
+        {
+          "$type": "AWS.Deploy.Common.Recipes.RequiredValidator, AWS.Deploy.Common",
```

Comment:
I'm not 100% onboarded with storing type definitions inside recipe. Ideally we shouldn't store any code. Perhaps some types that are mapped inside the code makes sense?

Reply:
changed to [use a factory](https://github.com/aws/aws-dotnet-deploy/pull/173/files?file-filters%5B%5D=.cs&file-filters%5B%5D=.csproj&file-filters%5B%5D=.json&file-filters%5B%5D=.recipe#diff-2850c845b3c55efd77c439f397b67fb21a97dbf651966d421ea85dce1f9d7e5eR28) instead of relying on deserializer